erigon-lib/kv/mdbx/kv_mdbx.go

```diff
@@ -61,53 +61,60 @@ func WithChaindataTables(defaultBuckets kv.TableCfg) kv.TableCfg {
 
 // handles background process of periodically flushing commits to disk
 type PeriodicFlusher struct {
-	env              *mdbx.Env // mdbx environment to flush to
-	opts             MdbxOpts
-	ticker           *time.Ticker  // set ticker
-	syncPeriod       time.Duration // how often to flush
-	quitFlushingChan chan struct{} // to signal end of flushing
-	closed           atomic.Bool
+	env        *mdbx.Env // mdbx environment to flush to
+	opts       MdbxOpts
+	ticker     *time.Ticker  // set ticker
+	syncPeriod time.Duration // how often to flush
+	closed     atomic.Bool
+	doneWg     sync.WaitGroup
 }
 
 func newPeriodicFlusher(env *mdbx.Env, opts MdbxOpts, syncPeriod time.Duration) *PeriodicFlusher {
 	return &PeriodicFlusher{
-		env:              env,
-		opts:             opts,
-		ticker:           time.NewTicker(syncPeriod),
-		syncPeriod:       syncPeriod,
-		quitFlushingChan: make(chan struct{}),
+		env:        env,
+		opts:       opts,
+		ticker:     time.NewTicker(syncPeriod),
+		syncPeriod: syncPeriod,
+		doneWg:     sync.WaitGroup{},
 	}
 }
 
-func (flusher *PeriodicFlusher) Close() {
+func (flusher *PeriodicFlusher) shutdown() {
+	// flusher.lock.Lock()
+	// defer flusher.lock.Unlock()
 	swapped := flusher.closed.CompareAndSwap(false, true)
 	if !swapped {
 		return
 	}
 	if flusher.ticker != nil {
 		flusher.ticker.Stop() // Stop the ticker
 	}
-	close(flusher.quitFlushingChan) //  close channel to signal quit
+}
+
+func (flusher *PeriodicFlusher) Close() {
+	flusher.shutdown()
+	flusher.doneWg.Wait()
```

Comment:
WaitGroup - must guarantee that “nothing will happen after shutdown” or “shutdown will happen after all other things”. 
Now “.Wait”.called after shutdown - doesn’t guarantee anything.

Reply:
You are waiting for shutdown to complete. Essentially you are waiting for the `for` loop to exit.